tianshou/data/buffer.py

```diff
@@ -359,6 +372,78 @@ def __getitem__(
             policy=self.get(index, "policy"),
         )
 
+    @classmethod
+    def _copy_to_hdf5(
```

Comment:
I'm considering if it is appropriate to move this functionality to `batch.py`. What do you think?

Reply:
Yes, sounds reasonable to move copy_from_hdf5 and copy_to_hdf5 to batch.py as they are not directly related to the replay buffer.

Reply:
I think you should have a look to how pickle works. Basically, the hdf5 data should be thé same than the pickle one, thus I think you should rather adapt the pickle interface to handle both hdf5 and usual pickle format. It is not Always possible, but since in this case you are playing with classes meant to deal with data, it should be doable fairly easy. At the end, it would be beneficial to pickle, which is likely to be more efficient if optimisez this way, but even more importantly easier to understand and maintain.

Reply:
Ok, I agree that unifying the data exposed to pickle and saved via HDF5 is a good idea. I implemented `__getstate__` and `__setstate__` in `ReplayBuffer` to work with a dictionary in which a `Batch` is represented by a dictionary (via `Batch.__getstate__`). `save_hdf5` is now based on the output of `__getstate__`. 
The main assumption when converting between this dictionary and HDF5 is that all entries which aren't dictionaries, numpy arrays or pytorch tensors can be saved as HDF5 attributes. According to the h5py documentation that means they have to be scalars (numpy arrays are already taken care of explicitly) and smaller than 64k.

Reply:
> The main assumption when converting between this dictionary and HDF5 is that all entries which aren't dictionaries

This is not necessary. Andy hierarchy is natively supported by hdf5 standard.

> According to the h5py documentation that means they have to be scalars (numpy arrays are already taken care of explicitly) 

What do you mean ? I dont get it.

> and smaller than 64k.

If such limitation actually existe, it is not part of hdf5 standard. Can you point me out where you found this information ?

Reply:
Hi, thank you for your reply. I was talking about attributes and not datasets. The information I am referring to can be found here: https://docs.h5py.org/en/stable/high/attr.html. It is also possible to store small numpy arrays as attributes but I explicitly save them in datasets. This is what I was trying to say.
Yes, it is not strictly necessary to make these assumptions. I can also implement something more general but this will increase the amount code needed. If you would prefer a solution that can deal with fairly general dictionaries and lists I can implement that.

Reply:
Do not make it overly general if not strictly necessary. I juste want to make sure your implementation is robust for the intended use, and flexible enough to be eventually extended if `ReplayBuffer` gets more complex in the future or new classes need to support from/to hdf5 interface.

Reply:
Also, note that the 64k limits is only a recommandation and not part of the standard. I'm not aware of any shortcoming of using attributs to store any kind of data, of any size.

Reply:
Thank you for pointing that out. I tried to address the robustness concerns by pickling objects that cannot be handled directly.